### PR TITLE
Security fix for docker packages

### PR DESCRIPTION
**Pull request by RemindD**
some important package version change:
- github.com/docker/docker from v24.0.9 to v27.1.1
- oras.land/oras-go from v1.2.4 to v1.2.6
- helm.sh/helm/v3 from v3.14.4 to v3.15.4
- k8s.io/client-go v0.29.0 from v0.29.0 to v0.30.3
- sigs.k8s.io/controller-runtime from v0.15.2 to v0.18.5

### CODE REVIEW

File: k8s/apis/config/v1/projectconfig_types.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ type ProjectConfig struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// ControllerManagerConfigurationSpec returns the contfigurations for controllers
-	cfg.ControllerManagerConfigurationSpec `json:",inline"`
+	// cfg.ControllerManagerConfigurationSpec `json:",inline"`
```
Comment (Haishi2016) [inline k8s/apis/config/v1/projectconfig_types.go:24]: I understand we are not actively using this, but why do we need to remove it?

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -89,6 +89,7 @@ jobs:
       run: |
         echo "TEST_KUBECTL:$TEST_KUBECTL TEST_MINIKUBE_ENABLED:$TEST_MINIKUBE_ENABLED TEST_K8S_STATE: $TEST_K8S_STATE TEST_CONFIGMAP: $TEST_CONFIGMAP"
         export REPOPATH="${{ github.workspace }}"
+        export DOCKER_API_VERSION=1.39
```
Comment (msftcoderdjw) [inline .github/workflows/go.yml:92]: Do we need this environment variable in symphony-api containers?

File: .github/workflows/go.yml [hunk 3]
```diff
@@ -33,7 +33,7 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v3
       with:
-        go-version: 1.22.3
+        go-version: 1.22.4
```
Comment (msftcoderdjw) [inline .github/workflows/go.yml:36]: test/localenv/README.md change the go installation here.

File: .github/workflows/integration.yml [hunk 4]
```diff
@@ -27,7 +27,7 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v3
       with:
-        go-version: 1.22.3
+        go-version: 1.22.4
```
Comment (msftcoderdjw) [inline .github/workflows/integration.yml:30]: Do we change in ADO repo as well?

File: api/pkg/apis/v1alpha1/providers/target/docker/docker.go [hunk 5]
```diff
@@ -240,7 +241,7 @@ func (i *DockerTargetProvider) Apply(ctx context.Context, deployment model.Deplo
 						return ret, err
 					}
 				}
-				err = cli.ContainerRemove(context.TODO(), component.Component.Name, types.ContainerRemoveOptions{})
+				err = cli.ContainerRemove(context.TODO(), component.Component.Name, container.RemoveOptions{})
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/docker/docker.go:244]: Can we pass ctx here?

File: api/pkg/apis/v1alpha1/providers/target/docker/docker.go [hunk 6]
```diff
@@ -290,7 +291,7 @@ func (i *DockerTargetProvider) Apply(ctx context.Context, deployment model.Deplo
 				return ret, err
 			}
 
-			if err = cli.ContainerStart(context.TODO(), container.ID, types.ContainerStartOptions{}); err != nil {
+			if err = cli.ContainerStart(context.TODO(), containerResponse.ID, container.StartOptions{}); err != nil {
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/docker/docker.go:294]: Can we pass ctx here?

File: api/Dockerfile [hunk 7]
```diff
@@ -4,7 +4,7 @@
 ## SPDX-License-Identifier: MIT
 ##
 
-ARG BUILD_BASE_IMAGE=golang:1.22.3-alpine
+ARG BUILD_BASE_IMAGE=golang:1.22.4-alpine
```
Comment (msftcoderdjw) [inline api/Dockerfile:7]: We need corresponding change in ADO repo.

File: k8s/apis/fabric/v1/webhook_suite_test.go [hunk 8]
```diff
@@ -83,12 +83,8 @@ var _ = BeforeSuite(func() {
 	// start webhook server using Manager
 	webhookInstallOptions := &testEnv.WebhookInstallOptions
 	mgr, err := ctrl.NewManager(cfg, ctrl.Options{
-		Scheme:             scheme,
```
Comment (msftcoderdjw) [inline k8s/apis/fabric/v1/webhook_suite_test.go:86]: Is this because controller-runtime package upgrade?

File: k8s/main.go [hunk 9]
```diff
@@ -204,22 +200,13 @@ func main() {
...
+
 	options := ctrl.Options{
 		Scheme:                 scheme,
-		MetricsBindAddress:     metricsAddr,
-		Port:                   9443,
 		HealthProbeBindAddress: probeAddr,
 		LeaderElection:         enableLeaderElection,
 		LeaderElectionID:       "33405cb8.symphony",
 	}
-	if configFile != "" {
```
Comment (msftcoderdjw) [inline k8s/main.go:216]: What is the new requirement of using controller config file in newer version controller-runtime?

File: .github/workflows/go.yml (hunk 3 above)
Comment (RemindD) [reply .github/workflows/go.yml:36]: I think already changed go version in test/localenv/README?

File: k8s/apis/fabric/v1/webhook_suite_test.go (hunk 8 above)
Comment (RemindD) [reply k8s/apis/fabric/v1/webhook_suite_test.go:86]: Yes

File: k8s/main.go (hunk 9 above)
Comment (RemindD) [reply k8s/main.go:216]: We need to implement our own config reader if needed. Reference: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/config/v1alpha1

File: api/Dockerfile (hunk 7 above)
Comment (RemindD) [reply api/Dockerfile:7]: This part will be done when adding commits to ADO repo.

File: .github/workflows/go.yml (hunk 2 above)
Comment (RemindD) [reply .github/workflows/go.yml]: discussed offline. We don't want to set the environment variable in API container.

**Review by msftcoderdjw — APPROVED** [review]

**Review by Haishi2016 — APPROVED** [review]

**Pull request merged**